src/common/sourceMaps/sourceMapResolutionUtils.ts

```diff
@@ -0,0 +1,182 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import * as utils from '../../common/urlUtils';
+import * as path from 'path';
+import { PathMapping } from '../../configuration';
+import { URL } from 'url';
+import { properJoin, fixDriveLetterAndSlashes, properResolve } from '../pathUtils';
+import { LogTag } from '../logging';
+import { logger } from '../logging/logger';
+import { exists } from '../fsUtils';
+import { filterObject } from '../objUtils';
+
+export function getFullSourceEntry(sourceRoot: string | undefined, sourcePath: string): string {
+  if (!sourceRoot) {
+    return sourcePath;
+  }
+
+  if (!sourceRoot.endsWith('/')) {
+    sourceRoot += '/';
+  }
+
+  return sourceRoot + sourcePath;
+}
+
+/**
+ * Gets the best source root out of the set of path mappings.
+ */
+export async function getComputedSourceRoot(
+  sourceRoot: string,
+  generatedPath: string,
+  pathMapping: PathMapping,
+  resolver: PathMappingResolver,
+): Promise<string> {
+  generatedPath = utils.fileUrlToAbsolutePath(generatedPath) || generatedPath;
+
+  let absSourceRoot: string;
+  if (sourceRoot) {
+    if (utils.isFileUrl(sourceRoot)) {
+      // sourceRoot points to a local path like "file:///c:/project/src", make it an absolute path
+      absSourceRoot = utils.fileUrlToAbsolutePath(sourceRoot);
+    } else if (utils.isAbsolute(sourceRoot)) {
+      // sourceRoot is like "/src", should be like http://localhost/src, resolve to a local path using pathMaping.
+      // If path mappings do not apply (e.g. node), assume that sourceRoot is actually a local absolute path.
+      // Technically not valid but it's easy to end up with paths like this.
+      absSourceRoot = (await resolver(sourceRoot, pathMapping)) || sourceRoot;
+
+      // If no pathMapping (node), use sourceRoot as is.
+      // But we also should handle an absolute sourceRoot for chrome? Does CDT handle that? No it does not, it interprets it as "localhost/full path here"
+    } else if (path.isAbsolute(generatedPath)) {
+      // sourceRoot is like "src" or "../src", relative to the script
+      absSourceRoot = resolveRelativeToFile(generatedPath, sourceRoot);
+    } else {
+      // generatedPath is a URL so runtime script is not on disk, resolve the sourceRoot location on disk.
```

Comment:
Can this branch happen now? It looks like generatedPath might always be an absolute path?

Reply:
We process the `sourceRoot` from the sourcemap like how we did before. The sourcemap spec is rather loose definition and I haven't personally seen sourceRoots that are URLs, but it's a code path in chrome-debug-core.